### PR TITLE
Ben/interface ros v0

### PR DESCRIPTION
**Pull request by bdschnapp**
created simple backend for the tablet interface using ros bridge. 

Known issue: launch file does not work, and rosbridgde server must be launched seperatly from ros_bridge_interface ros_bridge_interface.py

### CODE REVIEW

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: what does this mean? Is it the topic will control the screen changes?

File: nightingale_interface/screens/screen_config.py
Comment (FelixKHWong) [inline nightingale_interface/screens/screen_config.py]: What will these numbers represent? i was thinking that the variables "CURRENT_ACTION" and "PENDING_ACTION" would be the strings that we can check and also show on screen for more information. Do you think we should keep the integers and have a lookup table for strings if we need to show the action on screen?

File: ros_bridge_interface/package.xml
Comment (FelixKHWong) [inline ros_bridge_interface/package.xml]: replace with your email

File: ros_bridge_interface/package.xml
Comment (FelixKHWong) [inline ros_bridge_interface/package.xml]: MIT

File: ros_bridge_interface/package.xml
Comment (FelixKHWong) [inline ros_bridge_interface/package.xml]: do we need to add rosbridge here as a dependency to run?

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: do we want this print statement?

File: ros_bridge_interface/launch/interface.launch
Comment (FelixKHWong) [inline ros_bridge_interface/launch/interface.launch]: are these arguments default? Or will we need to change the port later?

**Review by FelixKHWong — CHANGES_REQUESTED** [review]
Can you look at the previous PR here https://github.com/nightingale-project/nightingale/pull/16 ? I think I won't approve this until the previous PR gets merged in

File: ros_bridge_interface/launch/interface.launch
Comment (bdschnapp) [reply ros_bridge_interface/launch/interface.launch]: This is the port that is set in MovoConfig. I believe the default is 9090. I would like to make this value a parameter and controlled from one spot eventually but didn't have time yet.

File: nightingale_interface/main.py
Comment (bdschnapp) [reply nightingale_interface/main.py]: good catch, that was for debugging

File: ros_bridge_interface/package.xml
Comment (bdschnapp) [reply ros_bridge_interface/package.xml]: nope, rosbridge server runs in it's own package and this package in just subscribing to the topic that rosbridge is publishing to

File: nightingale_interface/screens/screen_config.py
Comment (bdschnapp) [reply nightingale_interface/screens/screen_config.py]: My understanding of your code was that these were supposed to be enums, sorry if I missunderstood. It makes the most sense to use enums when sending data through

File: nightingale_interface/main.py
Comment (bdschnapp) [reply nightingale_interface/main.py]: yes this is the screen change topic

File: dev_notes.txt
Comment (FelixKHWong) [inline dev_notes.txt]: Create a new README file for this. I can probably do this on the previous PR

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: Why are the task and the args(I'm assuming data for each data) split up? Wouldn't it be best if the task is a single object which has fields containing the task ID and also fields with data for that task?

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: what do these action and interface topics do? I was thinking we would subscribe to a single status topic and make decisions based on the message recieved

File: ros_bridge_interface/src/ros_bridge_interface.py
Comment (FelixKHWong) [inline ros_bridge_interface/src/ros_bridge_interface.py]: Can we change the topic the interface node subscribes to to "nightingale_ui/user_input"? I think this makes more sense since it is listening to messages the tablet is sending for when the patient wants the robot to do something

File: ros_bridge_interface/src/ros_bridge_interface.py
Comment (FelixKHWong) [inline ros_bridge_interface/src/ros_bridge_interface.py]: Can we change the interface node published topic to "nightingale_ui/robot_status"? I think this makes more sense since the interface is relaying changes in the robot status for the tablet to reach to by changing the screen.

File: ros_bridge_interface/src/ros_bridge_interface.py
Comment (FelixKHWong) [inline ros_bridge_interface/src/ros_bridge_interface.py]: I see this is the callback for the interface to react to the data sent from the tablet. Where does this send the data back to the mission planner? Will there be a service called on the interface node which kicks off the publisher to send the status to the tablet and then the tablet responses with user input and then input is then sent back as a response for the service?

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py:66]: Now that we've discussed the system as sending a status to the UI and the UI sending the user input data back, is this "queue" functionality required? i feel like the mission planner will always send the next status when it is available and will not end up queuing multiple tasks to do at a time.

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: requested topic name change above

File: nightingale_interface/main.py
Comment (FelixKHWong) [inline nightingale_interface/main.py]: Should now be based on the "status" received

**Review by FelixKHWong — CHANGES_REQUESTED** [review]
I can also go onto your branch and make the changes if that's ok with you

File: nightingale_interface/main.py
Comment (bdschnapp) [reply nightingale_interface/main.py]: thats a good point, this should be a single variable instead of three and just use a dict to pass everything in to it

File: nightingale_interface/main.py
Comment (bdschnapp) [reply nightingale_interface/main.py:66]: This is more for blocking tasks to keep the touch screen unblocked rather than to actually queue tasks. This allows the execution of blocking tasks from a seperate thread

File: ros_bridge_interface/src/ros_bridge_interface.py
Comment (bdschnapp) [reply ros_bridge_interface/src/ros_bridge_interface.py]: The links between rb interface and mission planner need to be implemented. Services, actions, and topics that communicate with mission planner should be defined here, and the data recieved in this callback should be processed and forwarded to the correct service/action/topic

File: nightingale_interface/screens/screen_config.py
Comment (FelixKHWong) [reply nightingale_interface/screens/screen_config.py]: We'll use enums and then convert to strings later if needed

File: nightingale_interface/main.py
Comment (bdschnapp) [inline nightingale_interface/main.py]: "if func_args:" ? I don't see args defined here and I think this will cause a run time error

File: nightingale_ros_bridge/launch/interface.launch
Comment (bdschnapp) [inline nightingale_ros_bridge/launch/interface.launch:1]: Has this launch file been tested yet? Last time I used it it didn't seem to work and I had to launch the nodes individually

File: nightingale_ros_bridge/src/nightingale_ros_bridge/bridge_interface_config.py
Comment (bdschnapp) [inline nightingale_ros_bridge/src/nightingale_ros_bridge/bridge_interface_config.py]: Let's talk about this in person, I understand why this was done this way, but I don't like it and am having a hard time explaining why I don't like it

**Review by bdschnapp — COMMENTED** [review]
Looks good to me. Only the first comment needs to be fixed before merging. The other comments can be discussed later

File: nightingale_interface/main.py
Comment (FelixKHWong) [reply nightingale_interface/main.py:66]: I think we should remove this now that the Mission planner is doing the actual computation and the tablet is just an input interface. What do you think?

File: nightingale_ros_bridge/src/nightingale_ros_bridge/bridge_interface_config.py
Comment (FelixKHWong) [reply nightingale_ros_bridge/src/nightingale_ros_bridge/bridge_interface_config.py]: Yeah is it because it doesn't start at 1? I was concerned there would be mistakes made from overlap with the Robotstates. Having these start at 0 also makes sense since they come from different class objects

File: nightingale_interface/main.py
Comment (FelixKHWong) [reply nightingale_interface/main.py]: oh yeah this is a typo good catch. Do you think we should just remove this queue task stuff? It seems like the mission planner will be doing the task queue stuff and that the tablet is just being used as an interface

File: nightingale_ros_bridge/launch/interface.launch
Comment (FelixKHWong) [reply nightingale_ros_bridge/launch/interface.launch:1]: I think so, I'm running "roslaunch nightingale_ros_bridge interface.launch" and getting things communicating as you saw today

File: nightingale_interface/main.py
Comment (bdschnapp) [inline nightingale_interface/main.py]: This is dangerous code, any value could get put into self.ros_set_screen and if it is not a screen name the app will crash. Use something like: 
`if self.ros_set_screen and self.get_screen(self.ros_set_screen):`
`    self.root.current = self.ros_set_screen`
`    self.ros_set_screen = None`

the get_screen method is from the screen_wrapper class

File: nightingale_interface/main.py
Comment (bdschnapp) [reply nightingale_interface/main.py:66]: I disagree. Lets talk in person about this, it has made me think of more things that need to be implemented specifically related to the estop

File: nightingale_interface/screens/home_screen.py
Comment (bdschnapp) [inline nightingale_interface/screens/home_screen.py]: This is a good temporary fix, eventaully we can package the app to run as an executable though and the imports will be handeled in the build process. https://kivy.org/doc/stable/guide/packaging-windows.html

**Review by bdschnapp — COMMENTED** [review]
looks good! Only one comment needs to be fixed, I don't think this will need another review before merging

File: nightingale_interface/main.py
Comment (FelixKHWong) [reply nightingale_interface/main.py:66]: ok I'll leave it in for V0

**Review by FelixKHWong — DISMISSED** [review]
Should be ready to complete V0

File: nightingale_interface/main.py
Comment (younesr1) [inline nightingale_interface/main.py]: consider running a static analyzer like pylint https://pylint.pycqa.org/en/latest/user_guide/usage/run.html

**Review by bdschnapp — COMMENTED** [review]
Looks good!

File: nightingale_interface/main.py
Comment (FelixKHWong) [reply nightingale_interface/main.py]: I ran the linter and my code is pretty bad though with how Kivy works I think the score would be quite low regardless since it allows me to use variables in other classes which is normally not possible. It would take me too long to refactor and make things much better and I feel that the code is readable enough for now. I did use the linter to clean up some errors such as the import errors for now.

**Review by FelixKHWong — APPROVED** [review]

**Pull request merged**
